Add long-context evaluation benchmarks (LongBench v2, Babilong, InfiniteBench, Phonebook)

Mariani-code opened the pull request:
This PR Fixes #3224

# Add Long-Context Evaluation Benchmarks

## Summary

This PR adds four major long-context evaluation benchmarks to lm-evaluation-harness, enabling comprehensive evaluation of language models on tasks requiring processing of 1k to 10M+ tokens. These benchmarks are essential for evaluating the emerging capabilities of modern LLMs with extended context windows.

## Added Benchmarks

### 1. LongBench v2
- **Tasks**: 20 tasks across QA, summarization, code, and reasoning
- **Context Length**: 8k-2M tokens
- **Dataset**: `THUDM/LongBench` on HuggingFace Hub

### 2. Babilong
- **Tasks**: 20 reasoning tasks (qa1-qa20)
- **Context Length**: 1k-10M tokens
- **Dataset**: `booydar/babilong` on HuggingFace Hub

### 3. InfiniteBench
- **Tasks**: 12 tasks (retrieval, math, code, QA, dialogue)
- **Context Length**: 100k-1M+ tokens
- **Dataset**: `OpenBMB/InfiniteBench` on HuggingFace Hub

### 4. Phonebook Lookup
- **Tasks**: 7 variants testing position sensitivity
- **Context Length**: 1k-200k tokens
- **Dataset**: Synthetic generation (original dataset not on Hub)

## Implementation Details

### File Structure
```
lm_eval/tasks/
├── longbench_v2/
│   ├── README.md
│   ├── utils.py
│   ├── _longbench_v2.yaml
│   └── [20 task YAML files]
├── babilong/
│   ├── README.md
│   ├── utils.py
│   ├── _babilong.yaml
│   └── [20 task YAML files]
├── infinitebench/
│   ├── README.md
│   ├── utils.py
│   ├── _infinitebench.yaml
│   └── [12 task YAML files]
├── phonebook/
│   ├── README.md
│   ├── utils.py
│   ├── _phonebook.yaml
│   └── [7 task YAML files]
└── LONG_CONTEXT_BENCHMARKS.md
```

### Key Features
- **59 new tasks** total across all benchmarks
- **Comprehensive metrics**: F1, ROUGE, exact match, position-specific accuracy
- **Flexible configuration**: Support for various context lengths and task variants
- **Documentation**: Complete READMEs with citations and usage examples
- **Group configurations**: Easy evaluation of entire benchmarks

## Usage Examples

```bash
# Evaluate on all new benchmarks
lm_eval --model hf \
  --model_args pretrained=meta-llama/Llama-3-8B \
  --tasks longbench_v2,babilong,infinitebench,phonebook \
  --batch_size 1

# Evaluate specific context lengths
lm_eval --model hf \
  --model_args pretrained=gpt2 \
  --tasks phonebook_1k,phonebook_5k \
  --batch_size 1

# Evaluate a single benchmark
lm_eval --model hf \
  --model_args pretrained=mistral-7b \
  --tasks babilong \
  --batch_size 1
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=3256) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=3256) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EleutherAI/lm-evaluation-harness?pullRequest=3256) it.</sub>

StellaAthena commented on lm_eval/tasks/LONG_CONTEXT_BENCHMARKS.md:
I don't think we're going to want to include this file.

jannalulu commented:
Also babilong has up to qa20 and there's no test/train split

Mariani-code commented:
Hi @jannalulu,

Thanks for the comments! I've now added all 20 tasks to the implementation.

  To clarify the other points:

  1. Prompt Format: You're right that LongBench uses a standardized multiple-choice format. However, Babilong is a different benchmark with its own simple generative QA prompt format.
 
  2. Task Coverage (QA11-QA20):  The tasks qa11-qa20 exist but are structured differently in the dataset (as splits within length configs rather than separate configs). I've now added all 20 tasks with the appropriate loading mechanism.

  3. Train/Test Splits: The dataset doesn't use traditional train/test splits. The "splits" are based on context length buckets (0k, 1k, 2k...1M). This is by design for evaluating performance across different context lengths.
 
  4. LongBench v1 vs v2: You're correct about the similarities. While LongBench v2 has major changes, Babilong is actually a separate benchmark focused on long-context "needle-in-haystack" QA, so those changes don't directly affect this implementation.

jannalulu commented:
> Prompt Format: You're right that LongBench uses a standardized multiple-choice format. However, Babilong is a different benchmark with its own simple generative QA prompt format.

My point was that LongBench has different prompts in this implementation from the paper.

> Train/Test Splits: The dataset doesn't use traditional train/test splits. The "splits" are based on context length buckets (0k, 1k, 2k...1M). This is by design for evaluating performance across different context lengths.

Yes, the splits are technically "qa1, qa2, etc." and not context lengths; the subsets have the context lengths.

> LongBench v1 vs v2: You're correct about the similarities. While LongBench v2 has major changes, Babilong is actually a separate benchmark focused on long-context "needle-in-haystack" QA, so those changes don't directly affect this implementation.

What are the major changes in v1 vs. v2? I agree that Babilong is a separate benchmark. Also please see the prompts in the [babilong paper](https://arxiv.org/pdf/2406.10149).

foreverpiano commented:
LGTM. when does it get merged? @jannalulu

Mariani-code commented:
Modified dataset so the test cases should run now.
